File: Trees/Easy/array_to_BST.py
# https://practice.geeksforgeeks.org/problems/array-to-bst4443/1#
class Solution:
    ans = []
    def sortedArrayToBST(self, nums):
        self.ans = []
        self.solve(nums, 0, len(nums)-1)
        
        return self.ans
        
    def solve(self, nums, l, r):
        if(l > r):
            return
        
        m = (l + r) // 2
        self.ans.append(nums[m])
        
        self.solve(nums, l, m - 1)
        self.solve(nums, m + 1, r)

#{ 
#  Driver Code Starts
if __name__ == '__main__':
    T=int(input())
    for i in range(T):
        n = int(input())
        nums = list(map(int, input().split()))
        obj = Solution()
        ans = obj.sortedArrayToBST(nums)
        for _ in ans:
            print(_, end = " ")
        print()

# } Driver Code Ends

File: Trees/Easy/array_to_BST.cpp
// https://practice.geeksforgeeks.org/problems/array-to-bst4443/1#
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
class Solution {
public:
    vector<int> sortedArrayToBST(vector<int>& nums) {
        vector<int> ans;
        solve(nums, ans, 0, nums.size()-1);
        
        return ans;
    }
    
    void solve(vector<int>& nums, vector<int>& arr, int l, int r)
    {
        if(l > r) return;
        
        int m = (l + r) / 2;
        arr.push_back(nums[m]);
        solve(nums, arr, l, m -1);
        solve(nums, arr, m + 1, r);
    }
};
// { Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n;
		cin >> n;
		vector<int>nums(n);
		for(int i = 0; i < n; i++)cin >> nums[i];
		Solution obj;
		vector<int>ans = obj.sortedArrayToBST(nums);
		for(auto i: ans)
			cout << i <<" ";
		cout << "\n";
	}
	return 0;
}  // } Driver Code Ends